sdk/src/OpenTelemetry/Processor/Batch.hs

```diff
@@ -237,23 +284,58 @@ batchProcessor BatchTimeoutConfig {..} exporter = liftIO $ do
             case push span_ builder of
               Nothing -> (builder, True)
               Just b' -> (b', False)
-          when appendFailed $ void $ tryPutMVar workSignal Flush
-      , processorForceFlush = void $ tryPutMVar workSignal Flush
+          when appendFailed $ void $ atomically $ tryPutTMVar workSignal ()
+      , processorForceFlush = void $ atomically $ tryPutTMVar workSignal ()
       , -- TODO where to call restore, if anywhere?
-        processorShutdown = async $ mask $ \_restore -> do
-          -- flush remaining messages
-          void $ tryPutMVar workSignal Shutdown
-
-          shutdownResult <-
-            timeout (millisToMicros exportTimeoutMillis) $
-              wait worker
-          -- make sure the worker comes down
-          uninterruptibleCancel worker
-          -- TODO, not convinced we should shut down processor here
-
-          case shutdownResult of
-            Nothing -> pure ShutdownFailure
-            Just _ -> pure ShutdownSuccess
+        processorShutdown =
+          asyncWithUnmask $ \unmask -> unmask $ do
+            -- we call asyncWithUnmask here because the shutdown action is
+            -- likely to happen inside of a `finally` or `bracket`. the
+            -- @safe-exceptions@ pattern (followed by unliftio as well)
+            -- will use uninterruptibleMask in an exception cleanup. the
+            -- uninterruptibleMask state means that the `timeout` call
+            -- below will never exit, because `wait worker` will be in the
+            -- `uninterruptibleMasked` state, and the timeout async
+            -- exception will not be delivered.
+            --
+            -- see note [Unmasking Asyncs]
+            mask $ \_restore -> do
+              -- is it a little silly that we unmask and remask? seems
+              -- silly! but the `mask` here is doing an interruptible mask.
+              -- which means that async exceptions can still be delivered
+              -- if a process is blocking.
+
+              -- flush remaining messages and signal the worker to shutdown
+              void $ atomically $ putTMVar shutdownSignal ()
+
+              -- gracefully wait for the worker to stop. we may be in
+              -- a `bracket` or responding to an async exception, so we
+              -- must be very careful not to wait too long. the following
+              -- STM action will block, so we'll be susceptible to an async
+              -- exception.
+              delay <- registerDelay (millisToMicros exportTimeoutMillis)
+              shutdownResult <-
+                atomically $
+                  msum
+                    [ Just <$> waitCatchSTM worker
```

Comment:
I don't quite follow how `waitCatchSTM` does not block here, but STM blocking semantics seem overall quite very nonobvious to me. My understanding here is that `waitCatchSTM` will `retry` if it not done, then this code will run the check for the timer. Is that correct?

Reply:
`atomically` blocks if all of the actions `retry`. So this operation *does* block, and *can* throw an async exception, even inside of `mask`.

But, yeah, the logic is "First check `waitCatchSTM worker`. If that is not complete, then check the timer."

The GHC RTS is smart enough to not endlessly loop and continually retry, and will instead send a wakeup to any blocked `STM` action when any of the linked variables change. So `worker` here becomes a "watched" variable, as does `delay`. When `delay` gets updated by the other thread, then GHC will wake up this thread up, retry the transaction as a whole, and complete.

Now, it's possible we'd have this flow:

1. `atomically` starts and then blocks.
2. The timer thread finishes, and the `TVar Bool` becomes `True`.
3. *Before GHC can wake up the `atomically` thread*, the `waitCatchSTM` *also* finishes.
4. GHC wakes up `atomically`, which does `msum ...`.
5. `waitCatchSTM` is first, so it gets pulled.

But that's actually totally fine and what we want.